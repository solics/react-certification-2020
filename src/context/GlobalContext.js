import React, { createContext, useReducer } from 'react'
import { globalReducer, globalInitialState } from '../reducers'

const GlobalContext = createContext()

function GlobalContextProvider({ children }) {
	const [state, dispatch] = useReducer(globalReducer, globalInitialState)
	return (
		<GlobalContext.Provider value={[state, dispatch]}>{children}</GlobalContext.Provider>
	)
}

export { GlobalContext, GlobalContextProvider }
