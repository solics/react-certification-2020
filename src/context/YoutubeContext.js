import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const [loading, setLoading] = useState(false)
	const [loadingRelated, setLoadingRelated] = useState(false)
	const [errorMsg, setErrorMsg] = useState('')
	const [errorMsgRelated, setErrorMsgRelated] = useState('')
	const [videosRelated, setVideosRelated] = useState([])

	const search = term => {
		setLoading(true)
		fetch(`${SEARCH_URL}q=${term}`)
			.then(response => response.json())
			.then(data => {
				setLoading(false)
				setErrorMsg('')
				if (data?.items?.length) {
					setVideos(data.items)
					saveArrayOnLocalStorage('videos', data.items)
				} else {
					setVideos([])
					setErrorMsg(data.error.message)
				}
			})
			.catch(error => {
				setLoading(false)
				setErrorMsg(error.error.message)
				console.error(error)
			})
	}

	const getRelatedVideos = videoId => {
		setLoadingRelated(true)
		fetch(`${SEARCH_URL}relatedToVideoId=${videoId}`)
			.then(response => response.json())
			.then(data => {
				setLoadingRelated(false)
				setErrorMsgRelated('')
				if (data?.items?.length) {
					setVideosRelated(data.items)
				} else {
					setVideosRelated([])
					setErrorMsgRelated(data.error.message)
				}
			})
			.catch(error => {
				setLoadingRelated(false)
				setErrorMsgRelated(error.error.message)
				console.error(error)
			})
	}

	const defaultContext = {
		videos,
		videosRelated,
		search,
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
		getRelatedVideos,
		loading,
		loadingRelated,
		errorMsg,
		errorMsgRelated,
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
