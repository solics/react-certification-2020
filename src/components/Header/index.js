import React, { useContext, useEffect, useState } from 'react'
import styled from 'styled-components/macro'
import MENU_ICON from '../../assets/img/svg/menu.svg'
import USER_ICON from '../../assets/img/svg/user.svg'
import { GlobalContext } from '../../context/GlobalContext'
import SearchBox from '../SearchBox'
import SwitchTheme from '../SwitchTheme'
import THEMES from '../../themes'
import Login from '../../pages/Login'
import { loginApi } from '../../api/services'
import SideMenu from '../SideMenu'

const HeaderStyled = styled.header`
	padding: 20px;
	display: flex;
	justify-content: space-between;
	background-color: ${props => props.theme.headerBackground};
`
const MenuButton = styled.button`
	padding: 5px 10px 0 0px;
	background: none;
	border: none;
	cursor: pointer;
	outline: none;
	img {
		width: 70%;
	}
`
const HeaderLeft = styled.div`
	display: grid;
	grid-template-columns: 40px auto;
	width: 300px;
	@media only screen and (max-width: 768px) {
		width: 100%;
	}
`
const HeaderRight = styled.div`
	display: flex;
	position: relative;
	@media only screen and (max-width: 768px) {
		display: none;
	}
`
const Avatar = styled.div`
	width: 30px;
	height: 30px;
	border-radius: 50%;
	overflow: hidden;
	border: 2px solid white;
	cursor: pointer;
	img {
		width: 100%;
	}
`
const MenuLogIn = styled.div`
	position: absolute;
	bottom: -100%;
	right: 0;
	background-color: #ffffff;
	color: #000000;
	padding: 10px;
	cursor: pointer;
	&:hover {
		background-color: #cccccc;
	}
`

const Header = () => {
	const [state] = useContext(GlobalContext)
	const [showLoginBtn, setShowLoginBtn] = useState(false)
	const [showLoginModal, setShowLoginModal] = useState(false)
	const [avatarUrl, setAvatarUrl] = useState(USER_ICON)
	const [error, setError] = useState('')
	const [showSideMenu, setShowSideMenu] = useState(false)

	useEffect(() => {
		const clickOutside = e => {
			if (e.target?.id === 'modal') setShowLoginModal(false)
			else if (e.target?.id === 'modal-menu') setShowSideMenu(false)
			setError('')
		}
		document.addEventListener('click', clickOutside)
		return () => document.removeEventListener('click', clickOutside)
	}, [])

	const handleClickLogin = () => {
		setShowLoginBtn(false)
		setShowLoginModal(true)
	}

	const handleSubmitLogin = async ({ username, password }) => {
		try {
			const response = await loginApi({ username, password })
			setAvatarUrl(response.avatarUrl)
			setShowLoginModal(false)
		} catch (err) {
			console.log(err)
			setError('Username or password invalid')
		}
	}

	return (
		<>
			<HeaderStyled theme={THEMES[state.currentTheme]}>
				{showSideMenu && <SideMenu />}
				<HeaderLeft>
					<MenuButton onClick={() => setShowSideMenu(!showSideMenu)}>
						<img src={MENU_ICON} alt="menu" />
					</MenuButton>
					<SearchBox />
				</HeaderLeft>
				<HeaderRight>
					<SwitchTheme />
					<Avatar onClick={() => setShowLoginBtn(!showLoginBtn)}>
						<img src={avatarUrl} alt="avatar" />
					</Avatar>
					{showLoginBtn && (
						<MenuLogIn onClick={handleClickLogin}>
							<span>Log in</span>
						</MenuLogIn>
					)}
				</HeaderRight>
			</HeaderStyled>
			{showLoginModal && <Login handleSubmitLogin={handleSubmitLogin} error={error} />}
		</>
	)
}

export default Header
