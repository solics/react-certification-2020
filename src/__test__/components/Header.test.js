import React from 'react'
import { createMemoryHistory } from 'history'
import { Route, Router } from 'react-router'
import { render } from '@testing-library/react'
import Header from '../../components/Header'
import { GlobalContextProvider } from '../../context/GlobalContext'

const history = createMemoryHistory()
history.push('/video-detail/21321321')

export const renderWithRouter = Component =>
	render(
		<GlobalContextProvider>
			<Router history={history}>
				<Route>
					<Component />
				</Route>
			</Router>
		</GlobalContextProvider>
	)

describe('Header', () => {
	it('Header render with its context without crashing', () => {
		renderWithRouter(Header)
	})
	it('App Snapshot', () => {
		const { container } = renderWithRouter(Header)
		expect(container).toMatchSnapshot()
	})
})
