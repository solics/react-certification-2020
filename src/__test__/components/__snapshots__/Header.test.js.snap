// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header App Snapshot 1`] = `
<div>
  <header
    class="HeaderStyled-sc-1azer9k-0 fkXvoa"
  >
    <div
      class="HeaderLeft-sc-1azer9k-2 jIiuwi"
    >
      <button
        class="MenuButton-sc-1azer9k-1 iOJGbh"
      >
        <img
          alt="menu"
          src="menu.svg"
        />
      </button>
      <form>
        <input
          class="InputSearch-sc-9xs9n4-0 kxGDRJ"
          id="search-inp"
          placeholder="Search"
          type="text"
          value=""
        />
      </form>
    </div>
    <div
      class="HeaderRight-sc-1azer9k-3 hMBPQS"
    >
      <span
        class="IndicatorStyled-j2fz6n-1 fNjceH"
      >
        Dark Mode
      </span>
      <label
        class="SwitchToggle-j2fz6n-0 dMraMU"
      >
        <input
          type="checkbox"
        />
        <span />
      </label>
      <div
        class="Avatar-sc-1azer9k-4 esSGob"
      >
        <img
          alt="avatar"
          src="user.svg"
        />
      </div>
    </div>
  </header>
</div>
`;
